### ISSUE TITLE
Possible false positive in `discarded_futures` lint

### ISSUE THREAD

**Issue by bartekpacia** [dart-lang/linter#3739]
**Describe the issue**

I believe that the `discarded_futures` lint is too aggressive. Read on.

**To Reproduce**

Let's see the following simple Flutter code:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyWidget(),
    );
  }
}

class MyWidget extends StatelessWidget {
  const MyWidget({super.key});

  Future<String> createFuture(String someParam) async {
    // do work

    return Future.value('dash');
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: FutureBuilder<String>(
        future: createFuture('hi'),
        builder: (context, snapshot) {
          return const Text('just a placeholder');
        },
      ),
    );
  }
}

```

And let's enable `discarded_futures` lint in `analysis_options.yaml`:

```yaml
include: package:flutter_lints/flutter.yaml

linter:
  rules:
    discarded_futures: true
```

And let's run `dart analyze`:

```
$ dart analyze
Analyzing discarded_futures_lint_bug... 1.0s

   info • lib/main.dart:35:17 • Don't invoke asynchronous functions in non-async blocks. •
          discarded_futures

1 issue found.
```

That's how it looks like in my VSCode:
![Screenshot 2022-10-03 at 12 45 12 AM](https://user-images.githubusercontent.com/40357511/193479628-9cac27c3-f38f-4e88-95f3-da4296912107.png)

Ideally, please include an example.

**Expected behavior**

I'd expect no warnings. `FutureBuilder.future` expects a future, I'm giving it the future, we're happy, but there's `discarded_futures` yelling at us.

**Additional context**

This might be another issue, but I often see code like this:

```dart
BlocProvider<AuthCubit>(
  lazy: false,
  create: (_) => AuthCubit(
    loginClient: context.read(),
    forIntegrationtest: _forIntegrationTest,
  )..init(), // init is async
),
```

The `discarded_futures` lint also warns about that `..init()`, but we never care about it. Implementing the lint that is talked about in [this issue](https://github.com/dart-lang/sdk/issues/46218) would solve this problem.

**Comment by srawlins** [comment]
CC @pq who implemented `discared_futures` in https://github.com/dart-lang/linter/pull/3431

**Comment by jonmountjoy** [comment]
I've just added this lint to a fairly large code-base, and the only false-positive was the one highlighted in this issue.  The entire purpose of the `FutureBuilder` is to invoke an asynchronous function in a non-async block, so the warning of "Don't invoke asynchronous functions in non-async blocks" for this particular context of this lint, feels wrong.

**Comment by eernstg** [comment]
As far as I can see, `discarded_futures` does very nearly what the error message says, but it doesn't make any attempt to detect whether or not any futures are discarded. Here's an example error message:

```
   info • n009.dart:11:3 • Asynchronous function invoked in a non-'async'
          function. Try converting the enclosing function to be 'async' and
          then 'await' the future. • discarded_futures
```

It lints situations where a function returning a `Future<...>` is called in the body of a function whose return type isn't `Future<...>`. It doesn't care whether the caller or callee has an `async` body or not, but it is true that every function whose body is `async` must return a future.

This means that the lint allows functions returning a future to be synchronous and still call some `async` bodied functions:

```dart
Future<int> f() async => 1;

// Returns a future.
Future<int> g() {
  f(); // Calling an `async` function and discarding a future. No lint.
  return new Future<int>.value(0);
}

// Returns a non-future.
void h() {
  f(); // Lint!
}
```

I think the lint is working in a manner that makes sense: It is basically nudging developers in the direction of keeping all asynchronous computations in `async` function bodies. It is a simple test, and it yields advice which is probably a good habit in general.

It also makes sense that it isn't a core lint: There are lots of situations where it _does_ make sense that a synchronous function body calls a function that returns a future and then, say, stores that future or passes it as an actual argument to some other function. In other words, false positives are no surprise, and special application domains or coding styles may generate many false positives.

However, it should perhaps be renamed to something like `asynchronous_call_in_synchronous_function`, in order to avoid the misleading reference to 'discarded'?

**Comment by westito** [comment]
Renaming won't match the actual operation. When you return the future as parameter it is actually not called! So `asynchronous_call_in_synchronous_function` nether a good name. `discarded_futures` simply not working properly. Should not lint in this case. However, when it is encapsulated in try-catch it should lint because if the function not awaited, the try-catch will not catch. There is lint in DCM for this: https://dcm.dev/docs/rules/common/prefer-return-await. It should be linted in Dart linter too.

**Comment by andy-clapham** [comment]
I'm struggling with using this lint too - in theory it's so important it should be core. But in practice it doesn't seem to be handling the right thing.

Flutter's FutureBuilder is a classic case - the Future is simply not being discarded, it's being handled very nicely.

Synchronously taking a reference to a future in any code in a variable (sync or async) shoudn't be linted. Where the bugs occur is if the future is no longer referenced before it is considered completed via an await, precompleted, sync'd, or another mechanism like unawaited. 

e.g. this is fine

```dart

Future<void> doAsyncFuture() async {
  await Future<void>.delayed(const Duration(seconds: 1));
}

Future<void> nonAsyncFuture() {
  return Future.value(null);
}

void main() {
  final myFuture = doAsyncFuture();
  unawaited(myFuture);
  final useThisLater = nonAsyncFuture(); 
}

```

Whereas these are potential bugs, yet hugely common especially in Flutter callback handlers.

```dart
  Future<void> waitFor1sMaybe() async {
    doAsyncFuture();
  }

  final widget = Thing(
    onPressed: () async {
      doAsyncFuture();
      Navigator.pop(context);
    },
  );
```

Discarding _incomplete_ futures as they go out of scope is where the bugs happen. It can be considered separate from anything to do with async/await, except as await being one mechanism for ensuring a future is complete.

With deeper analysis, method-side opt out shouldn't be needed in the majority of cases, as the linter should see the future is being passed elsewhere or returned and not being discarded. But I suppose it's one way of systemizing the non-core unawaited library call into the linter without taking a dependency. Probably more relevant for dart-lang/sdk#46218

**Comment by gnprice** [comment]
It would be great to fix this lint rule (or make a new lint rule) to better live up to the rule's name: spotting futures that have been discarded, with some reasonable rate of false positives.

As is, the `discarded_futures` rule is incompatible with using some perfectly reasonable APIs like `Stream.fromFuture`, or Flutter's [`FutureBuilder`](https://main-api.flutter.dev/flutter/widgets/FutureBuilder-class.html). And though its documentation accurately describes what it does, that doesn't agree with its name. There seems to be a steady stream of reports in this tracker from people trying the rule and being surprised by that mismatch; some of the most recent are:
* https://github.com/dart-lang/linter/issues/5053
* https://github.com/dart-lang/linter/issues/4949
* https://github.com/dart-lang/linter/issues/4866
* https://github.com/dart-lang/linter/issues/4788
* https://github.com/dart-lang/linter/issues/4773
* as well as this thread.

---

One possible rule might be that if the future is (a) passed to some other method or (b) returned (or (c) is the argument to an `await`), then it's not discarded and doesn't cause a lint message. I think that would resolve all the examples I've seen, either in my own codebase or these threads.

To phrase it from the other direction, I believe that rule may be equivalent to saying — but I haven't tried to verify the equivalence, and could easily be missing some cases — that a future is "discarded" just if it's the top-level expression of an expression statement. I think that would also correctly handle all the examples I've seen, including both the "good" and "bad" examples in the handy comment at https://github.com/dart-lang/linter/issues/3739#issuecomment-1917438389 just above.
